src/api/server.rkt

```diff
@@ -53,75 +82,394 @@
                     #:timeline-disabled? [timeline-disabled? #f])
   (herbie-command command test seed pcontext profile? timeline-disabled?))
 
+(define (start-job-server)
+  (set! receptionist (make-receptionist)))
+
 ;; Starts a job for a given command object|
 (define (start-job command)
   (define job-id (compute-job-id command))
-  (if (already-computed? job-id) job-id (start-work command)))
+  (place-channel-put receptionist (list 'start receptionist command job-id))
+  (when verbose
+    (eprintf "Job ~a, Qed up for program: ~a\n" job-id (test-name (herbie-command-test command))))
+  job-id)
 
 (define (is-job-finished job-id)
-  (hash-ref *job-status* job-id #f))
+  (define-values (a b) (place-channel))
+  (place-channel-put receptionist (list 'check job-id b))
+  (when verbose
+    (eprintf "Checking if job is finished: ~a.\n" job-id))
+  (place-channel-get a))
+
+; verbose logging for debugging
+(define verbose #f) ; Maybe change to log-level and use 'verbose?
+(define (log msg)
+  (when verbose
+    ;; TODO fix string interpolation
+    (eprintf "~a\n" msg)))
 
 (define (wait-for-job job-id)
-  (if (already-computed? job-id) (hash-ref *completed-jobs* job-id) (internal-wait-for-job job-id)))
+  (define-values (a b) (place-channel))
+  (place-channel-put receptionist (list 'wait job-id b))
+  (define finished-result (place-channel-get a))
+  (when verbose
+    (eprintf "Done waiting for: ~a\n" job-id))
+  finished-result)
+
+(define (make-receptionist)
+  (place/context*
+   ch
+   #:parameters (*flags* *num-iterations*
+                         *num-points*
+                         *timeout*
+                         *reeval-pts*
+                         *node-limit*
+                         *max-find-range-depth*
+                         *pareto-mode*
+                         *platform-name*
+                         *loose-plugins*)
+   (parameterize ([current-error-port (open-output-nowhere)]) ; hide output
+     (load-herbie-plugins))
+   ; not sure if the above code is actaully needed.
+   (define completed-work (make-hash))
+   (define workers (make-hash))
+   (define waiting (make-hash))
+   (when verbose
+     (eprintf "Receptionist waiting for work.\n"))
+   (for ([i (in-naturals)])
+     ;  (eprintf "Receptionist msg ~a handled\n" i)
+     (match (place-channel-get ch)
+       ; Returns the current count of workers.
+       [(list 'count handler) (place-channel-put handler (hash-count workers))]
+       ; Retreive the improve results for results.json
+       [(list 'improve handler)
+        (define improved-list
+          (for/list ([(job-id result) completed-work]
```

Comment:
It's fine but I think it's better to wrap `for` targets with `in-hash`, `in-list`, etc. So here it'd be `(for/list ([(job-id result) (in-hash completed-work)])`

Reply:
It's faster but more importantly better error messages.